### ISSUE TITLE
[FR] Copy entities

### ISSUE THREAD

**Issue by Eneroth3** [SketchUp/api-issue-tracker#41]
One of the most common tasks when modeling is to copy entities using Move tool. This seemingly fundamental operation is quite hard to achieve via the API.

The recommended way of doing it today is probably to read all the data from thee entities, modify it and draw anew but it is quite tricky  and mundane to get all data, like welded curves or UV mapping.

In practise we often use a hack based on generation of a temporary group like this:

```ruby
def copy_entities(ent_ary, transformation)
  entities = ent_ary.first.parent.entities
  unless (ent_ary - entities.to_a).empty?
    raise ArgumentError, "All entities must be in the same entities collection."
  end
  
  # HACK: Create a temp group to be able to copy entities.
  # SketcUp crashes here if the entities isn't in the active drawing context.
  # Tested on SU2017 and SU2018.
  source_group = entities.add_group(ent_ary)
  
  entities.add_instance(source_group.definition, transformation).explode
  source_group.explode

  nil
end

# Move selected entities 5 m along red axis:
transformation = Geom::Transformation.new([5.m, 0, 0])
model = Sketchup.active_model
model.start_operation("Copy")
copy_entities(model.selection.to_a, transformation)
model.commit_operation
```

While this gets the job done, it is based on a hack that is discouraged in the documentation, and hard to figure out for new users.

I propose a sister method to Entities#transform_entities. This Entities#copy_entities would take the same arguments and behave the same, but create a copy instead of moving, similarly to how the move operation has a sister copy operation in the UI.

**Comment by Eneroth3** [comment]
Entities#copy_entities could return an array of all newly created entities, which may contain more elements than what was originally passed to the method , as edges and faces can have been split of by existing geometry.

**Comment by marvingreenberg** [comment]
Using the technique outlined above can also result in unexpected issues - when a face AND its attached edges are copied, edges that are part of an ArcCurve can lose their curve metadata.  E.g.

    copy_entities( [su_face] + su_face.edges, IDENTITY )   # using Eneroth3 method above 

This is possibly because the group.definition creation creates the boundary edges in the copied face, and the additional edge array segments _overwrite_ them.  In a newly provided Entities#copy_entities, either the implementation should ignore duplicate entities or entities that have been implicitly copied, or clearly document issues that may arise when copying.

**Comment by Eneroth3** [comment]
I updated my code example to take the transformation of the temp group into account (the temp group isn't formed with the identity transformation) and to raise and error if the entities passed are not in the active drawing context.

The suggested sister method #copy_entities should work on all entities collections, just as #transform_entities, not just the active one. I'm filing a new issue for #add_group causing SketchUp to crash.

@marvingreenberg 

I can't reproduce your issue with curves being exploded. The curves I've tested with keep being curves even after running it.

**Comment by marvingreenberg** [comment]
I uploaded a video to youtube.  My analysis was incorrect - it has nothing to do with passing the face and its edges.  Under some circumstances, this stripping of Arc data happens.  If you just draw a face, it seems to work fine to copy it.  But if you pull the face into a solid, and then copy just the face, I observe this issue, whether or not the call to  make the group is made from just the face, or the face and its edges.  See, https://youtu.be/Uy0O0x9pIR4

**Comment by DanRathbun** [comment]
> @Eneroth3  Entities#copy_entities could return an array of all newly created entities, which may contain more elements than what was originally passed to the method , as edges and faces can have been split of by existing geometry.

**Please DO return array of new elements !** 

Since the original entities will be passed into the method, we know coders may want references to the newly created geometry. (These new objects may need to be tagged with attributes, or their PIDs logged, etc. In some cases subsequent geometric operations will be performed upon these new objects immediately afterward.)

**Comment by DanRathbun** [comment]
I wonder how this could help reduce the (3 linked operation) rigmarole I went through in this example ...

https://forums.sketchup.com/t/code-undoable-place-component-movetool-clone/27193

**Comment by DanRathbun** [comment]
@Eneroth3 I think that the **transform** should be the 1st argument like the `Entities#transform_entities()` method has. This way the rest of the arguments can be an array or a list of variable length. Ie ...
```ruby
def copy_entities( transformation, *ent_ary )
```

**Comment by Eneroth3** [comment]
> I think that the transform should be the 1st argument like the Entities#transform_entities() method has.

Yup, they should definitely be consistent. Sorry my example code for the current work around/hack wasn't.

**Comment by thomthom** [comment]
Logged as SU-40234
